### ISSUE TITLE
Assign Users to Organizer

### ISSUE THREAD

**Issue by emergingdzns** [Attendize/Attendize#442]
My first client to spin up on Attendize has a need to be able to add users to the system and assign them to a particular organizer. We found the invite users feature and we can invite all the users we need to add, BUT in this organization, we need to be able to block users from browsing around different organizer accounts. Each user, if not an admin, needs to be able to log in and simply view the events inside their assigned Organizer "account".

Before I dive into tackling this myself, I would like to check and see if anyone has already done this and I've just missed it somehow.

My initial thought is to make this very simple and just add an organizer_id column to the users table and key that to the organizer table. Then in the login screen, if the user is assigned to an organizer, then it won't show the list of organizers, but would simply take them to their organizer. Then, block the list of organizations from showing up in the nav on the top right. Might be a few other places to add this rule that I haven't found yet, but this would be the start.

Alternatively, I could make it a bit more complex and have a pivot table for users_organizers and have it so that users can be assigned to multiple organizations. That would require making that login screen list the organizations based on the pivot table that they are assigned to. That top right nav would also have to change to show only those they have been assigned to.

The account management tool where you invite users would need changes as well to block assigned (non admin) users from inviting other users as well as having the ability for admin users to assign invited users to organizers.

For my client the former is the simplest and fastest approach, but trying to thing more long term for the community as a whole the  latter makes the most sense. Anyone have any thoughts or preferences on either approach? Or has someone already got a solution for this that I can implement?

**Comment by heilgar** [comment]
Hi, I looking, probably for the same feature, and thought about your example with pivot table for user_organizer. 
But after I start thinking about more complex way. would be great to add permissions for an action for users you described. For example like permissions it in laravel voyager.

FE: 
- New tab for organizer user management ( invite / assign event / permissions management )

 BE:
- user_organizer table
- event_user table (many to many) (cant have a look on db structure for now)
- permission_tree table 
- user_permissions table
- impersonate to event

Small plan :) 
What you think @emergingdzns ? Maybe we could collaborate somehow on that?

**Comment by emergingdzns** [comment]
Well my client is mainly focused on keeping users just locked to an organizer, but the did express an interest in possibly adding some level of permission. Things like:
canCreateTicket
canEditTicket
canDeleteTicket
canCreateEvent
canEditEvent
canDeleteEvent
canPublishEvent

I haven't thought through all the traits yet. But my plan was to do the user assign to organizer first. Then attempt to do permissions. I was planning to use one of the available role repos out there. I've used entrust before and really liked it. But it's old now so I'm not sure it's up to date enough. there's also laravel-permission and laravel-role. I need to look into them further. But I'd stick with one of these packages to provide the role/permission skeleton and just build off of that in the attendize system by applying the traits and rules. This may be getting too far ahead of myself though.

For the moment, once I get the organizer/user thing worked out, I'd probably just code this client's system for now to make it so that non-admin users (any user assigned to an organizer basically) can only READ for now. The only exceptions are the checkin updates and issuing refunds etc. No ability to create, update or delete events or tickets.

**Comment by emergingdzns** [comment]
To be honest, I'm not really sure. The only thing that is still kind of making me scratch my head is how to handle inviting a user and assigning them to an organizer at the same time. I'm thinking that the text field that we see now for entering an email address would need to be 1/2 as wide and then another field inline (drop down with multiselect if allowing more than one organizer) would need to be there. I'm not sure if a standard multiselect is good enough or if we should use some other library. If there is one already being used in Attendize then great we can use it, but otherwise, we'll need to find a different one like Chosen or something that we can style to fit the interface.

The same assignment multiselect should be on each user row that isn't an account owner. If not assigned it's an admin with global privileges. Once assigned it's locked. Only admins and account owner can access the entire Account Settings modal.

I'm open to suggestions on a good multi-select tool that would fit the style of Attendize interface.

**Comment by emergingdzns** [comment]
@heilgar I have issued a pull request (https://github.com/Attendize/Attendize/pull/447) that now includes the user functionality discussed above. There's no permissions functionality yet, but will be something I'll take on before too long.

**Comment by MelonSmasher** [comment]
Along the lines of this, would it be possible to assign Stripe keys to an organizer? I'm in a situation where multiple departments within our organization need to be able to sell tickets independently off each other and collect funds independently.

**Comment by vitorpamplona** [comment]
@MelonSmasher You might be looking for Feature Request #542

**Comment by philip-loyd** [comment]
Was this issue resolved? I was wrestling with this issue of permissions connected to organizers as well, but it occurred to me there's actually an issue with not having an *ADMIN panel* to manage accounts appropriately. This issue would easily be solved if you use the app as intended... Each **Account** has one or more **Organisers** where their events are hosted. The **Account** owner can invite sub-users, but they will be separate from all other accounts. Simple role change for sub-users would restrict users from changing any account settings... only an account owner should be able to update payments and critical info associated to the event, but sub-users should have access to everything else.

As for managing all accounts and their sub-users on a high-level, I have been testing Laravel Nova and think it could be a great addition to your project! You would need to buy a license, but it seems to fit in quite nice for a quick and dirty admin panel you can customize and improve upon. Thoughts?

**Comment by justynpride** [comment]
@philip-loyd have you integrated with Laravel Nova or just investigating. There are a couple of GitHub projects such as https://github.com/spatie/laravel-permission and https://github.com/jeremykenedy/laravel-users though they may not be as powerful as Nova.

**Comment by philip-loyd** [comment]
@justynpride thank you for sharing! We did install Nova, but apart from providing the dashboard to view/manage users, accounts, organisers, events, etc; it doesn't make the job of implementing user permissions on the front end as simple as I thought. I'll have to try those projects you sent - they look perfect!

**Comment by justynpride** [comment]
@philip-loyd Hole the links turn out to be helpful for you. I’d definitely be interested to hear how it goes. I could see how a good user management feature could become essential for other features for the system. I’d personally love some one to be able to book tickets, create a user account and to then be able to log in and view past orders and edit ‘live’ orders.

**Comment by philip-loyd** [comment]
100% agree, although I'm planning to completely remove certain account features that don't fit into my business model. For example, I'm removing the ability for anyone to add their own Stripe/Paypal account (at least until Stripe Connect is added). All users will be using the same bank account, and new account owners will need to go through a very selective approval process to be activated (managed through Nova admin panel). I do think a Stripe Connect many-to-many platform payment model would work very well once everything is in place... I'm currently paying out users through PayPal as a workaround. Here's a rough outline of what I have in mind for updates...

Add Frontend roles: 
- Owner (Full access to Organisers assigned to by admin)
- Staff (Extend access to any or all features through checkboxes) 
- Guest (View-only access to event management panel)
- Attendee (Can buy tickets, but unable to view any event management features) 

Add Nova Backend roles: 
- Admin (Full access site-wide)
- Developer (Full access with a few exceptions)
- Reporter (View only admin access)
- Contributor (Blog writer, editor) *fyi, we added a blog*
- Support (Limited access admin, semi-Owner manage access to assist users)

There's still a lot more to work through to make all this happen, so any feedback or suggestions you have would be incredibly helpful! My goal is to mirror Eventbrite's structure as much as possible. https://www.eventbrite.com/support/articles/en_US/Multi_Group_How_To/how-to-manage-roles-and-permissions-in-eventbrite-music?lg=en_US

**Comment by justynpride** [comment]
Glad to hear those links are of help. It might be worth checking out laravelvoyager.com as well. 

Your plans sound exciting. I definitely agree on Stripe Connect, as it will solve a lot of problems and create much more opportunities. I'd definitely be interesting in helping with testing if it helps you, or just the final version! Would your 'Attendee' role be able to view their order history, and also be able to edit 'live' orders? By editing I'm thinking of adding, swapping, or cancelling tickets, or updating details on tickets?

**Comment by justynpride** [comment]
@emergingdzns Just wondering where you got to with your Users to Organisers feature? Were you able to complete?
